File: src/main/java/com/salon24/crawler/SiteCrawler.java
package com.salon24.crawler;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

@Component
class SiteCrawler {
    private final SiteProcessor siteProcessor;
    private final UrlExtractor urlExtractor;

    private final Set<String> visited = new HashSet<>();

    @Autowired
    SiteCrawler(SiteProcessor siteProcessor, UrlExtractor urlExtractor) {
        this.siteProcessor = siteProcessor;
        this.urlExtractor = urlExtractor;
    }

    void crawl(String siteUrl) {
        Document document = downloadSite(siteUrl);
        PartitionedUrls urls = urlExtractor.getUrlsFromSite(document);

        siteProcessor.process(new SiteInfo(siteUrl, document, urls));
        visited.add(siteUrl);

        urls.getAll().stream()
                .filter(url -> !visited.contains(url))
                .forEach(this::crawl);
    }

    private Document downloadSite(String url) {
        try {
            return Jsoup.connect(url).get();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/com/salon24/crawler/SiteCrawler.kt
package com.salon24.crawler

import com.salon24.processor.SiteProcessorFactory
import org.jsoup.Jsoup
import org.springframework.stereotype.Component

/**
 * WERY NAIW IMPLEMENTEJSZON. SZUD BI DAN ON FREDS.
 */
@Component
class SiteCrawler(private val urlExtractor: UrlExtractor, private val siteClasifier: SiteClasifier) {
    private val processed = mutableSetOf<String>()

    fun crawl(url: String) {
        val site = extractSiteInfo(url)

        val siteProcessor = SiteProcessorFactory.fromType(site.type)
        siteProcessor.process(site)

        processed.add(url)

        urlExtractor.extractSalon24UrlsFromSite(site.document)
                .filterNot { processed.contains(it) }
                .forEach { crawl(it) }
    }

    private fun extractSiteInfo(url: String) =
            SiteInfo(
                    url = url,
                    type = siteClasifier.getSiteTypeByUrl(url),
                    document = Jsoup.connect(url).get()
            )
}